lisa/runners/lisa_runner.py

```diff
@@ -20,298 +21,384 @@
-                results, use_new_environment=True, environment_status=environment.status
-            ):
-                self._check_cancel()
-                if new_env_result.check_environment(environment, True):
-                    self._run_suite(
-                        environment=environment, case_results=[new_env_result]
-                    )
-                    break
-
-        # grouped test results by test suite.
-        grouped_cases: List[TestResult] = []
-        current_test_suite: Optional[TestSuiteMetadata] = None
-        for test_result in self._get_can_run_results(
-            results, use_new_environment=False, environment_status=environment.status
-        ):
-            self._check_cancel()
-            if test_result.check_environment(environment, True):
-                if (
-                    test_result.runtime_data.metadata.suite != current_test_suite
-                    and grouped_cases
-                ):
-                    # run last batch cases
-                    self._run_suite(environment=environment, case_results=grouped_cases)
-                    grouped_cases = []
-
-                # append new test cases
-                current_test_suite = test_result.runtime_data.metadata.suite
-                grouped_cases.append(test_result)
-
-        if grouped_cases:
-            self._run_suite(environment=environment, case_results=grouped_cases)
-
-    def _run_suite(
-        self, environment: Environment, case_results: List[TestResult]
-    ) -> None:
-
-        assert case_results
-        suite_metadata = case_results[0].runtime_data.metadata.suite
+        assert test_results
+        suite_metadata = test_results[0].runtime_data.metadata.suite
         test_suite: TestSuite = suite_metadata.test_class(
             suite_metadata,
         )
-        test_suite.start(environment=environment, case_results=case_results)
+        test_suite.start(environment=environment, case_results=test_results)
+
+    def _delete_environment_task(
+        self, environment: Environment, test_results: List[TestResult] = []
+    ) -> None:
+        """
+        May be called async
+        """
+        # the predefined environment shouldn't be deleted, because it
+        # serves all test cases.
+        if (
+            environment.status
+            in [
+                EnvironmentStatus.Deployed,
+                EnvironmentStatus.Connected,
+            ]
+        ) or (
+            environment.status == EnvironmentStatus.Prepared and environment.is_in_use
+        ):
+            self.platform.delete_environment(environment)
+        else:
+            environment.status = EnvironmentStatus.Deleted
+
+    def _get_same_priority_results(
+        self, test_results: List[TestResult]
+    ) -> List[TestResult]:
+        if not test_results:
+            return []
+
+        current_priority = test_results[0].runtime_data.metadata.priority
+        test_results = [
```

Comment:
We can have a loop here to avoid the code duplication

Reply:
It should be a pattern like `do...while` to perform first run before loop. But I don't know how to do it in Python.

Reply:
while True and break. We have to deal with it.

Reply:
ok, will fix like this.

Reply:
I see, missed one more step in my look up. All good, disregard :)